File: src/main/java/me/anonymoussoftware/vacancymanager/ui/panels/VacancyDescriptionPanel.java
package me.anonymoussoftware.vacancymanager.ui.panels;

import java.awt.BorderLayout;
import java.awt.EventQueue;

import javax.swing.BoxLayout;
import javax.swing.JButton;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JTextPane;

import me.anonymoussoftware.vacancymanager.App;
import me.anonymoussoftware.vacancymanager.VacancyManager;
import me.anonymoussoftware.vacancymanager.VacancyManager.VacancyDescriptionRequestListener;
import me.anonymoussoftware.vacancymanager.VacancyManager.VacancySelectionListener;
import me.anonymoussoftware.vacancymanager.model.Employer;
import me.anonymoussoftware.vacancymanager.model.Salary;
import me.anonymoussoftware.vacancymanager.model.Snippet;
import me.anonymoussoftware.vacancymanager.model.Vacancy;
import me.anonymoussoftware.vacancymanager.model.aggregated.AggregatedVacancy;

@SuppressWarnings("serial")
public class VacancyDescriptionPanel extends JPanel implements VacancySelectionListener, VacancyDescriptionRequestListener {

    private final JTextPane vacancyDescriptionTextArea;

    private final VacancyManager vacancyManager = App.getBean(VacancyManager.class);

    private final JButton banVacancyButton;

    private final JButton banEmployerButton;

    private final JButton requestDescriptionButton;

    public VacancyDescriptionPanel() {
        setLayout(new BorderLayout());

        this.vacancyDescriptionTextArea = new JTextPane();
        this.vacancyDescriptionTextArea.setContentType("text/html");
        add(new JScrollPane(this.vacancyDescriptionTextArea), BorderLayout.CENTER);

        JPanel buttonPanel = new JPanel();
        buttonPanel.setLayout(new BoxLayout(buttonPanel, BoxLayout.Y_AXIS));

        this.banVacancyButton = new JButton("Ban vacancy");
        this.banVacancyButton.setEnabled(false);
        this.banVacancyButton.addActionListener(e -> this.vacancyManager.banSelectedVacancy());
        buttonPanel.add(this.banVacancyButton);

        this.banEmployerButton = new JButton("Ban company");
        this.banEmployerButton.addActionListener(e -> this.vacancyManager.banSelectedVacancyEmployer());
        this.banEmployerButton.setEnabled(false);
        buttonPanel.add(banEmployerButton);

        this.requestDescriptionButton = new JButton("Request vacancy description");
        this.requestDescriptionButton.addActionListener(e -> this.vacancyManager.requestSelectedVacancyDescription());
        this.requestDescriptionButton.setEnabled(false);
        buttonPanel.add(requestDescriptionButton);

        add(buttonPanel, BorderLayout.EAST);
    }

    @Override
    public void addNotify() {
        super.addNotify();
        this.vacancyManager.addVacancySelectionListener(this);
        this.vacancyManager.addVacancyDescriptionRequestListener(this);
    }

    @Override
    public void removeNotify() {
        this.vacancyManager.removeVacancyDescriptionRequestListener(this);
        this.vacancyManager.removeVacancySelectionListener(this);
        super.removeNotify();
    }

    @Override
    public void onVacancySelection() {
        AggregatedVacancy aggregatedVacancy = this.vacancyManager.getSelectedVacancy();
        StringBuilder description = new StringBuilder();
        if (aggregatedVacancy != null) {
            Vacancy vacancy = aggregatedVacancy.getVacancy();
            if (vacancy != null) {
                description.append(String.format("<b>%s</b>", vacancy.getName()));
                description.append("<br/>");
                description.append("URL: ");
                description.append(vacancy.getUrl());
                description.append("<br/>");
                Salary salary = vacancy.getSalary();
                if (salary != null) {
                    description.append(String.format("<b>%s</b>", salary.toString()));
                    description.append("<br/>");
                }
                description.append("<br/>");
                Employer employer = vacancy.getEmployer();
                if (employer != null) {
                    description.append(String.format("<b>%s</b>", employer.getName()));
                    description.append("<br/>");
                    description.append("URL: ");
                    description.append(employer.getUrl());
                    description.append("<br/>");
                    description.append("<br/>");
                }
                Snippet snippet = vacancy.getSnippet();
                if (snippet != null) {
                    description.append(snippet.toString());
                    description.append("<br/>");
                    description.append("<br/>");
                }
                String vacancyDescription = vacancy.getDescription();
                if (vacancyDescription != null && !vacancyDescription.isEmpty()) {
                    description.append("<b>Detailed description: </b>");
                    description.append("<br/>");
                    description.append(vacancyDescription);
                }
            }
        }
        EventQueue.invokeLater(() -> {
            String vacancyDescription = description.toString().trim();
            if (vacancyDescription.isEmpty()) {
                vacancyDescription = "<br/>";
            }
            this.vacancyDescriptionTextArea.setText(vacancyDescription);
            this.banVacancyButton.setEnabled(aggregatedVacancy != null && !aggregatedVacancy.getVacancy().isBanned());
            this.banEmployerButton
                    .setEnabled(aggregatedVacancy != null && !aggregatedVacancy.getVacancy().getEmployer().isBanned());
            this.requestDescriptionButton.setEnabled(aggregatedVacancy != null //
                    && (aggregatedVacancy.getVacancy().getDescription() == null
                            || aggregatedVacancy.getVacancy().getDescription().isEmpty()));
        });
    }

    @Override
    public void onVacancyDescriptionRequestStart(String url) {
        EventQueue.invokeLater(() -> {
            this.banVacancyButton.setEnabled(false);
            this.banEmployerButton.setEnabled(false);
            this.requestDescriptionButton.setEnabled(false);
        });
    }

    @Override
    public void onVacancyDescriptionRequestFinish() {
        onVacancySelection();
    }
}

File: src/main/kotlin/me/anonymoussoftware/vacancymanager/ui/panels/VacancyDescriptionPanel.kt
package me.anonymoussoftware.vacancymanager.ui.panels

import me.anonymoussoftware.vacancymanager.App
import me.anonymoussoftware.vacancymanager.VacancyManager
import me.anonymoussoftware.vacancymanager.VacancyManager.VacancyDescriptionRequestListener
import me.anonymoussoftware.vacancymanager.VacancyManager.VacancySelectionListener
import java.awt.BorderLayout
import java.awt.EventQueue
import javax.swing.*

class VacancyDescriptionPanel : JPanel(), VacancySelectionListener, VacancyDescriptionRequestListener {

    private val vacancyDescriptionTextArea: JTextPane

    private val vacancyManager = App.getBean(VacancyManager::class.java)

    private val banVacancyButton: JButton

    private val banEmployerButton: JButton

    private val requestDescriptionButton: JButton

    init {
        layout = BorderLayout()

        this.vacancyDescriptionTextArea = JTextPane()
        this.vacancyDescriptionTextArea.contentType = "text/html"
        add(JScrollPane(this.vacancyDescriptionTextArea), BorderLayout.CENTER)

        val buttonPanel = JPanel()
        buttonPanel.layout = BoxLayout(buttonPanel, BoxLayout.Y_AXIS)

        this.banVacancyButton = JButton("Ban vacancy")
        this.banVacancyButton.isEnabled = false
        this.banVacancyButton.addActionListener { _ -> this.vacancyManager.banSelectedVacancy() }
        buttonPanel.add(this.banVacancyButton)

        this.banEmployerButton = JButton("Ban company")
        this.banEmployerButton.addActionListener { _ -> this.vacancyManager.banSelectedVacancyEmployer() }
        this.banEmployerButton.isEnabled = false
        buttonPanel.add(banEmployerButton)

        this.requestDescriptionButton = JButton("Request vacancy description")
        this.requestDescriptionButton.addActionListener { _ -> this.vacancyManager.requestSelectedVacancyDescription() }
        this.requestDescriptionButton.isEnabled = false
        buttonPanel.add(requestDescriptionButton)

        add(buttonPanel, BorderLayout.EAST)
    }

    override fun addNotify() {
        super.addNotify()
        this.vacancyManager.addVacancySelectionListener(this)
        this.vacancyManager.addVacancyDescriptionRequestListener(this)
    }

    override fun removeNotify() {
        this.vacancyManager.removeVacancyDescriptionRequestListener(this)
        this.vacancyManager.removeVacancySelectionListener(this)
        super.removeNotify()
    }

    override fun onVacancySelection() {
        val aggregatedVacancy = this.vacancyManager.selectedVacancy
        val description = StringBuilder()
        if (aggregatedVacancy != null) {
            val vacancy = aggregatedVacancy.vacancy
            if (vacancy != null) {
                description.append(
                    """
                    <b>${vacancy.name}</b>
                    <br/>
                    URL: ${vacancy.url}
                    <br/>
                    """
                )
                val salary = vacancy.salary
                if (salary != null) {
                    description.append("<b>$salary</b><br/>")
                }
                description.append(
                    """
                    <br/>
                    <b>${vacancy.employer.name}</b>
                    <br/>
                    URL: ${vacancy.employer.url}
                    <br/>
                    <br/>
                    """
                )
                val snippet = vacancy.snippet
                if (snippet != null) {
                    description.append("$snippet<br/><br/>")
                }
                val vacancyDescription = vacancy.description
                if (vacancyDescription != null && vacancyDescription.isNotEmpty()) {
                    description.append("<b>Detailed description: </b><br/>$vacancyDescription")
                }
            }
        }
        EventQueue.invokeLater {
            var vacancyDescription = description.toString().trim { it <= ' ' }
            if (vacancyDescription.isEmpty()) {
                vacancyDescription = "<br/>"
            }
            this.vacancyDescriptionTextArea.text = vacancyDescription
            this.banVacancyButton.isEnabled = aggregatedVacancy != null && !aggregatedVacancy.vacancy!!.isBanned
            this.banEmployerButton.isEnabled =
                aggregatedVacancy != null && !aggregatedVacancy.vacancy!!.employer.isBanned
            this.requestDescriptionButton.isEnabled = (aggregatedVacancy != null //
                    && (aggregatedVacancy.vacancy!!.description == null || aggregatedVacancy.vacancy.description!!.isEmpty()))
        }
    }

    override fun onVacancyDescriptionRequestStart(url: String) {
        EventQueue.invokeLater {
            this.banVacancyButton.isEnabled = false
            this.banEmployerButton.isEnabled = false
            this.requestDescriptionButton.isEnabled = false
        }
    }

    override fun onVacancyDescriptionRequestFinish() {
        onVacancySelection()
    }
}
